runtime/src/iree/hal/drivers/hip/hip_device.c

```diff
@@ -89,6 +89,151 @@ typedef struct iree_hal_hip_device_t {
 } iree_hal_hip_device_t;
 
 static const iree_hal_device_vtable_t iree_hal_hip_device_vtable;
+static const iree_hal_deferred_work_queue_device_interface_vtable_t
+    iree_hal_hip_deferred_work_queue_device_interface_vtable;
+
+// We put a hipEvent_t into a void*.
+static_assert(sizeof(hipEvent_t) <= sizeof(void*), "Unexpected event size");
+typedef struct iree_hal_hip_deferred_work_queue_device_interface_t {
+  iree_hal_deferred_work_queue_device_interface_t base;
+  iree_hal_device_t* device;
+  hipDevice_t hip_device;
+  hipCtx_t hip_context;
+  hipStream_t dispatch_hip_stream;
+  iree_allocator_t host_allocator;
+  const iree_hal_hip_dynamic_symbols_t* hip_symbols;
+} iree_hal_hip_deferred_work_queue_device_interface_t;
+
+void iree_hal_hip_deferred_work_queue_symbol_table_destroy(
+    iree_hal_deferred_work_queue_device_interface_t* symbol_table) {
+  iree_hal_hip_deferred_work_queue_device_interface_t* table =
+      (iree_hal_hip_deferred_work_queue_device_interface_t*)(symbol_table);
+  iree_allocator_free(table->host_allocator, table);
+}
+
+iree_status_t iree_hal_hip_deferred_work_queue_symbol_table_bind_to_thread(
+    iree_hal_deferred_work_queue_device_interface_t* symbol_table) {
+  iree_hal_hip_deferred_work_queue_device_interface_t* table =
+      (iree_hal_hip_deferred_work_queue_device_interface_t*)(symbol_table);
+  return IREE_HIP_RESULT_TO_STATUS(table->hip_symbols,
+                                   hipCtxSetCurrent(table->hip_context),
+                                   "hipCtxSetCurrent");
+}
+
+iree_status_t iree_hal_hip_deferred_work_queue_symbol_table_wait_native_event(
+    iree_hal_deferred_work_queue_device_interface_t* symbol_table,
+    void* event) {
+  iree_hal_hip_deferred_work_queue_device_interface_t* table =
+      (iree_hal_hip_deferred_work_queue_device_interface_t*)(symbol_table);
+  hipEvent_t hip_event = (hipEvent_t)(event);
+  return IREE_HIP_RESULT_TO_STATUS(
+      table->hip_symbols,
+      hipStreamWaitEvent(table->dispatch_hip_stream, hip_event, 0),
+      "hipStreamWaitEvent");
+}
+
+iree_status_t iree_hal_hip_deferred_work_queue_symbol_table_create_native_event(
+    iree_hal_deferred_work_queue_device_interface_t* symbol_table,
+    void** out_event) {
+  iree_hal_hip_deferred_work_queue_device_interface_t* table =
+      (iree_hal_hip_deferred_work_queue_device_interface_t*)(symbol_table);
+  hipEvent_t* out = (hipEvent_t*)(out_event);
```

Comment:
```suggestion
  hipEvent_t* out = (hipEvent_t*)out_event;
```
no need for extra parens (needed in macros, but not here)